video-streaming-app-with-directus-and-sveltekit/index.md

```diff
@@ -0,0 +1,387 @@
+{/if}
+```
+Directus stores file metadata in the `directus_files` collection. When using file fields in other collections, Directus creates a one-to-many relationship. To include file data when fetching items, you use dot notation in the fields parameter, like `thumbnail.*` and `video_file.*`. This tells Directus to populate the file information from the `directus_files` collection.
+
+![Video Listing](<Screenshot 2024-07-04 at 11.53.25.png>)
+
+## Video Player Page
+Install the Svelte video player component to play the videos with the command.
+
+```shell
+npm install svelte-video-player
+```
+
+Update your `services/index.ts` file to add new functions that will fetch a video by its ID and update the `videos` collection to increment the video's views field.
+```ts
++
+// your other imports
+import { readItems, readItem, updateItem } from "@directus/sdk";
+
+export async function getVideo(id: string): Promise<Video> {
+  const directus = getDirectusClient();
+  const response = await directus.request(
+    readItem("videos", id, {
+      fields: ["*", "thumbnail.*", "video_file.*"],
+    })
+  );
+  return response as Video;
+}
+
+export async function incrementViews(id: string) {
+  const directus = getDirectusClient();
+  const video = await directus.request(readItem("videos", id));
+  await directus.request(
+    updateItem("videos", id, { views: parseInt(video.views || 0) + 1 })
+  );
+}
+```
+
+Create a nested route in your **routes** folder in the format `video/[id]/+page.svelte` to create a page to play selected videos. Update this file with the following code:
+
+```svelte
+<script lang="ts">
+  import { page } from "$app/stores";
+  import { getVideo, incrementViews } from "$lib/services";
+  import VideoPlayer from "svelte-video-player";
+  import type { Video } from "$lib/types";
+
+  let video: Video | null = null;
+
+  $: id = $page.params.id;
+  $: if (id) {
+    getVideo(id)
+      .then((v: Video) => {
+        video = v;
+        incrementViews(id);
+      })
+      .catch((error) => {
+        console.error("Error fetching video:", error);
+      });
+  }
+</script>
+
+{#if video}
+  <h1>{video.title}</h1>
+  <p>
+    {video.views} views • {new Date(video.upload_date).toLocaleDateString()}
+  </p>
+  <VideoPlayer
+    poster={`${import.meta.env.VITE_DIRECTUS_URL}/assets/${video.thumbnail.id}`}
+    source={`${import.meta.env.VITE_DIRECTUS_URL}/assets/${video.video_file.id}`}
+  />
+  <h2>Description</h2>
+  <p>{video.description}</p>
+  <h2>Tags</h2>
+  <div class="tags">
+    {#each video.tags as tag}
+      <span class="tag">{tag}</span>
+    {/each}
+  </div>
+{:else}
+  <p>Loading...</p>
+{/if}
+
+<style>
+  .tags {
+    display: flex;
+    flex-wrap: wrap;
+    gap: 0.5rem;
+  }
+  .tag {
+    background-color: #eee;
+    padding: 0.25rem 0.5rem;
+    border-radius: 0.25rem;
+  }
+</style>
+```
+Now click on any of the videos to stream it. 
+
+![Video Player](<Screenshot 2024-07-04 at 12.27.56.png>)
```

Comment:
````suggestion
## Creating the Video Player Page

Update your `services/index.ts` file to add new functions that will fetch a video by its ID and update the `videos` collection to increment the video's views field.

```ts
// your other imports
import { readItems, readItem, updateItem } from "@directus/sdk";

export async function getVideo(id: string): Promise<Video> {
  const directus = getDirectusClient();
  const response = await directus.request(
    readItem("videos", id, {
      fields: ["*", "thumbnail.*", "video_file.*"],
    })
  );
  return response as Video;
}

export async function incrementViews(id: string) {
  const directus = getDirectusClient();
  const video = await directus.request(readItem("videos", id));
  await directus.request(
    updateItem("videos", id, { views: parseInt(video.views || 0) + 1 })
  );
}
```

Create a nested route in your **routes** folder in the format `video/[id]/+page.svelte` to create a page to play selected videos. Update this file with the following code:

```svelte
<script lang="ts">
  import { page } from "$app/stores";
  import { getVideo, incrementViews } from "$lib/services";
  import VideoPlayer from "svelte-video-player";
  import type { Video } from "$lib/types";

  let video: Video | null = null;

  $: id = $page.params.id;
  $: if (id) {
    getVideo(id)
      .then((v: Video) => {
        video = v;
        incrementViews(id);
      })
      .catch((error) => {
        console.error("Error fetching video:", error);
      });
  }
</script>

{#if video}
  <h1>{video.title}</h1>
  <p>
    {video.views} views • {new Date(video.upload_date).toLocaleDateString()}
  </p>
  <VideoPlayer
    poster={`${import.meta.env.VITE_DIRECTUS_URL}/assets/${video.thumbnail.id}`}
    source={`${import.meta.env.VITE_DIRECTUS_URL}/assets/${video.video_file.id}`}
  />
  <h2>Description</h2>
  <p>{video.description}</p>
  <h2>Tags</h2>
  <div class="tags">
    {#each video.tags as tag}
      <span class="tag">{tag}</span>
    {/each}
  </div>
{:else}
  <p>Loading...</p>
{/if}

<style>
  .tags { display: flex; flex-wrap: wrap; gap: 0.5rem; }
  .tag { background-color: #eee; padding: 0.25rem 0.5rem; border-radius: 0.25rem; }
</style>
```
Now click on any of the videos to stream it. 

![Video Player](<Screenshot 2024-07-04 at 12.27.56.png>)
````

Sorry to be annoying but can you double check the syntax in the `video/[id]/+page.svelte` `<script>`. I'm not a huge Svelte user, but looks a bit funky so a double check is always best.

Reply:
The syntax is correct for a Svelte application :)